sources/api/certdog/src/main.rs

```diff
@@ -0,0 +1,401 @@
+        let data = bundle.data.clone().unwrap_or_default();
+
+        // Empty data means the certificate bundle was disabled in the API
+        if data.trim() == "" {
+            debug!("Found empty bundle: {}", name);
+            continue;
+        }
+
+        let name = name.as_ref();
+        let decoded = base64::decode(data.as_bytes()).context(error::Base64Decode { name })?;
+        // Each record in the API could include one or more certificates
+        let mut pems = pems_from_iter(x509_parser::pem::Pem::iter_from_buffer(&decoded))?;
+
+        // `trusted` defaults to false if not set in the API record
+        if bundle.trusted.unwrap_or(false) {
+            trusted.append(&mut pems);
+        } else {
+            distrusted.append(&mut pems);
+        }
+    }
+
+    Ok((trusted, distrusted))
+}
+
+/// Updates the trusted certificates store, removing the distrusted certificates
+/// from the final bundle
+fn update_trusted_store<P>(
+    (mut trusted, distrusted): (Vec<x509_parser::pem::Pem>, Vec<x509_parser::pem::Pem>),
+    trusted_store: P,
+    source_bundle: P,
+    add_comment: bool,
+) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    let source_bundle = source_bundle.as_ref();
+    let trusted_store = trusted_store.as_ref();
+
+    // The default bundle includes the certificates shipped with the OS
+    let default_bundle = fs::File::open(source_bundle).context(error::ReadFile {
+        path: source_bundle,
+    })?;
+    let reader = BufReader::new(default_bundle);
+    // Initialize trusted bundle with the certificates shipped with the OS
+    let mut trusted_bundle = pems_from_iter(x509_parser::pem::Pem::iter_from_reader(reader))?;
+
+    // Add additional trusted certificates
+    trusted_bundle.append(&mut trusted);
+    // Remove any distrusted certificate
+    trusted_bundle.retain(|pem| !distrusted.contains(pem));
+
+    // Write a PEM formatted bundle from trusted certificates
+    fs::write(
+        &trusted_store,
+        pems_to_string(&trusted_bundle, add_comment)?,
+    )
+    .context(error::UpdateTrustStore)?;
+
+    Ok(())
+}
+
+/// Returns a list with Pem objects from a PemIterator
+fn pems_from_iter<R>(iter: x509_parser::pem::PemIterator<R>) -> Result<Vec<x509_parser::pem::Pem>>
+where
+    R: BufRead + Seek,
+{
+    let mut vec: Vec<x509_parser::pem::Pem> = Vec::new();
+    for pem in iter {
+        let pem = pem.context(error::ParsePEM)?;
+        vec.push(pem);
+    }
+    Ok(vec)
+}
+
+/// Concatenates all the PEM objects as a single PEM bundle
+fn pems_to_string(pems: &Vec<x509_parser::pem::Pem>, add_comment: bool) -> Result<String> {
+    let mut out = String::new();
+
+    for pem in pems {
+        writeln!(out, "{}", pem_to_string(pem, add_comment)?).context(error::WritePEMString)?;
+    }
+
+    Ok(out)
+}
+
+/// Transforms a PEM object into a PEM formatted string
+fn pem_to_string(pem: &x509_parser::pem::Pem, add_comment: bool) -> Result<String> {
```

Comment:
This and `pems_to_string` also seem test-able, but would be covered by `update_trusted_store`.